README.md
# flutter_datepicker_single

[![Build Status](https://travis-ci.com/adaptant-labs/flutter_datepicker.svg?branch=master)](https://travis-ci.com/adaptant-labs/flutter_datepicker#)
[![Pub](https://img.shields.io/pub/v/flutter_datepicker_single.svg)](https://pub.dartlang.org/packages/flutter_datepicker_single)

A flutter plugin for individual invocation of Date Pickers (Year, Month, Day)

![Example App Walkthrough](https://raw.githubusercontent.com/adaptant-labs/flutter_datepicker/master/datepicker-walkthrough.gif)

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/adaptant-labs/flutter_datepicker/issues

## License

Licensed under the terms of the Apache 2.0 license, the full version of which can be found in the
[LICENSE](https://raw.githubusercontent.com/adaptant-labs/flutter_datepicker/master/LICENSE) file included in the distribution.


CHANGELOG.md
## 0.0.1+1

* Minor cleanups recommended by pub.

## 0.0.1

* Initial release.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_datepicker_single/flutter_datepicker_single.dart';
import 'package:intl/intl.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter DatePicker Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter DatePicker Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final DateTime now = DateTime.now();
  DateTime _selectedDate;
  FlutterDatePickerMode _mode;
  int _result;

  @override
  void initState() {
    super.initState();
    _result = 0;
    _selectedDate = now;
  }

  // Demonstrates the dialog flow and Y-M-D cascading using showDatePicker,
  // for comparison.
  Future<DateTime> selectDateWithDatePicker(BuildContext context) async {
    return await showDatePicker(
      context: context,
      initialDate: _selectedDate,
      initialDatePickerMode: DatePickerMode.year,
      firstDate: DateTime(0),
      lastDate: DateTime(9999),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: IntrinsicWidth(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              Text("Selected Date - Day: " +
                  _selectedDate.day.toString() +
                  ", Month: " +
                  _selectedDate.month.toString() +
                  ", Year: " +
                  _selectedDate.year.toString()),
              const SizedBox(height: 24.0),
              _mode != null
                  ? Text("Selection Granularity: " + describeEnum(_mode),
                      style: TextStyle(color: Colors.red))
                  : Container(),
              _result != 0
                  ? Text("Selection Result: " + _result.toString(),
                      style: TextStyle(color: Colors.red))
                  : Container(),
              const SizedBox(height: 48.0),
              RaisedButton(
                  color: Colors.blue,
                  child: const Text("Select date with showDatePicker",
                      style: TextStyle(color: Colors.white)),
                  onPressed: () async {
                    DateTime selected = await selectDateWithDatePicker(context);

                    if (selected != null) {
                      setState(() {
                        _selectedDate = selected;
                        _mode = null;
                        _result = 0;
                      });
                    }
                  }),
              RaisedButton(
                child: const Text("Select a year with YearPicker"),
                onPressed: () async {
                  DateTime selected = await showYearPicker(
                    context: context,
                    initialDate: _selectedDate,
                  );

                  if (selected != null) {
                    setState(() {
                      _selectedDate = selected;
                      _mode = FlutterDatePickerMode.year;
                      _result = selected.year;
                    });
                  }
                },
              ),
              RaisedButton(
                child: const Text("Select a month with MonthPicker"),
                onPressed: () async {
                  DateTime selected = await showMonthPicker(
                    context: context,
                    initialDate: _selectedDate,
                    title: Text(DateFormat.MMMM().format(_selectedDate) +
                        " " +
                        _selectedDate.year.toString()),
                  );

                  if (selected != null) {
                    setState(() {
                      _selectedDate = selected;
                      _mode = FlutterDatePickerMode.month;
                      _result = selected.month;
                    });
                  }
                },
              ),
              RaisedButton(
                child: const Text("Select a day with DayPicker"),
                onPressed: () async {
                  DateTime selected = await showDayPicker(
                    context: context,
                    initialDate: _selectedDate,
                  );

                  if (selected != null) {
                    setState(() {
                      _selectedDate = selected;
                      _mode = FlutterDatePickerMode.day;
                      _result = selected.day;
                    });
                  }
                },
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/flutter_datepicker.dart
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart';

/// Type of date picker for [FlutterDatePicker] to use.
///
/// This is primarily used as an internal mechanism for the various accessors
/// ([showYearPicker], [showMonthPicker], [showDayPicker]) to signal their
/// required picker type, but is exposed to support direct instantiation of
/// [FlutterDatePicker], where desired.
enum FlutterDatePickerMode {
  /// Show a date picker for choosing a day.
  day,

  /// Show a date picker for choosing a month.
  month,

  /// Show a date picker for choosing a year.
  year,
}

/// Select a year by calling the [YearPicker] class directly.
Future<DateTime> showYearPicker({
  @required BuildContext context,
  @required DateTime initialDate,
  DateTime firstDate,
  DateTime lastDate,
  Widget title,
  EdgeInsetsGeometry titlePadding,
}) async {
  assert(context != null);
  assert(initialDate  != null);

  return await showDialog(
      context: context,
      builder: (context) => FlutterDatePicker(
          mode: FlutterDatePickerMode.year,
          selectedDate: initialDate,
          firstDate: firstDate ?? DateTime(0),
          lastDate: lastDate ?? DateTime(9999),
          title: title,
          titlePadding: titlePadding,
      ),
  );
}

/// Select a month by calling the [MonthPicker] class directly.
Future<DateTime> showMonthPicker({
  @required BuildContext context,
  @required DateTime initialDate,
  DateTime firstDate,
  DateTime lastDate,
  Widget title,
  EdgeInsetsGeometry titlePadding,
}) async {
  assert(context != null);
  assert(initialDate  != null);

  return await showDialog(
    context: context,
    builder: (context) => FlutterDatePicker(
        mode: FlutterDatePickerMode.month,
        selectedDate: initialDate,
        firstDate: firstDate ?? DateTime(0),
        lastDate: lastDate ?? DateTime(9999),
        title: title,
        titlePadding: titlePadding,
    ),
  );
}

/// Select a day by calling the [DayPicker] class directly.
Future<DateTime> showDayPicker({
  @required BuildContext context,
  @required DateTime initialDate,
  DateTime firstDate,
  DateTime lastDate,
  Widget title,
  EdgeInsetsGeometry titlePadding,
}) async {
  assert(context != null);
  assert(initialDate  != null);

  return await showDialog(
    context: context,
    builder: (context) => FlutterDatePicker(
        mode: FlutterDatePickerMode.day,
        selectedDate: initialDate,
        firstDate: firstDate ?? DateTime(0),
        lastDate: lastDate ?? DateTime(9999),
        title: title,
        titlePadding: titlePadding,
    ),
  );
}

/// A non-cascading [SimpleDialog] view of a [FlutterDatePickerMode] picker.
///
/// In general this class does not need to be used directly, and should
/// be primarily accessed through the [showYearPicker], [showMonthPicker]
/// and [showDayPicker] convenience wrappers.
class FlutterDatePicker extends StatefulWidget {
  final FlutterDatePickerMode mode;
  final EdgeInsetsGeometry titlePadding;
  final Widget title;
  final DateTime selectedDate;
  final DateTime firstDate;
  final DateTime lastDate;

  FlutterDatePicker({
    Key key,
    @required this.mode,
    this.title,
    this.titlePadding,
    @required this.selectedDate,
    @required this.firstDate,
    @required this.lastDate
  }) : assert(!firstDate.isAfter(lastDate)),
       assert(selectedDate.isAfter(firstDate) &&
              selectedDate.isBefore(lastDate)),
       super(key: key);

  @override
  _FlutterDatePickerState createState() => _FlutterDatePickerState();
}

class _FlutterDatePickerState extends State<FlutterDatePicker> {
  final GlobalKey _pickerKey = GlobalKey();
  DateTime _selectedDate;

  @override
  void initState() {
    super.initState();
    _selectedDate = widget.selectedDate;
  }

  Widget _defaultPickerHeader(BuildContext context) {
    return Container(
        padding: EdgeInsets.all(20.0),
        child: Text("Select a " + describeEnum(widget.mode),
            style: Theme
                .of(context)
                .textTheme
                .title
                .copyWith(color: Colors.white)),
        color: Theme
            .of(context)
            .primaryColor);
  }

  @override
  Widget build(BuildContext context) {
    return SimpleDialog(
        titlePadding: widget.titlePadding ??
              widget.title == null ? EdgeInsets.all(0) : EdgeInsets.all(20.0),
        title: widget.title ?? _defaultPickerHeader(context),
        children: <Widget>[
          Container(
            height: 300,
            width: double.maxFinite,
            child:
              widget.mode == FlutterDatePickerMode.year ?
                YearPicker(
                  key: _pickerKey,
                  firstDate: widget.firstDate,
                  selectedDate: _selectedDate,
                  lastDate: widget.lastDate,
                  onChanged: (DateTime selected) {
                    setState(() {
                      _selectedDate = selected;
                    });
                  },
                ) :
              widget.mode == FlutterDatePickerMode.month ?
                MonthPicker(
                  key: _pickerKey,
                  firstDate: widget.firstDate,
                  selectedDate: _selectedDate,
                  lastDate: widget.lastDate,
                  onChanged: (DateTime selected) {
                    setState(() {
                      _selectedDate = selected;
                    });
                  },
                ) :
                DayPicker(
                  key: _pickerKey,
                  displayedMonth: widget.selectedDate,
                  firstDate: widget.firstDate,
                  selectedDate: _selectedDate,
                  currentDate: DateTime.now(),
                  lastDate: widget.lastDate,
                  onChanged: (DateTime selected) {
                    setState(() {
                      _selectedDate = selected;
                    });
                  },
                ),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.end,
            children: <Widget>[
              FlatButton(
                onPressed: () {
                  Navigator.of(context).pop();
                },
                child: Text('CANCEL',
                    textAlign: TextAlign.end,
                    style: TextStyle(
                        color: Theme.of(context).accentColor,
                        fontWeight: FontWeight.bold)),
              ),
              FlatButton(
                onPressed: () {
                  Navigator.of(context).pop(_selectedDate);
                },
                child: Text('OK',
                    textAlign: TextAlign.end,
                    style: TextStyle(
                        color: Theme.of(context).accentColor,
                        fontWeight: FontWeight.bold)),
              ),
            ],
          ),
        ]);
  }
}

lib/flutter_datepicker_single.dart
import 'package:flutter/material.dart';
import 'package:flutter/foundation.dart';

/// Type of date picker for [FlutterDatePicker] to use.
///
/// This is primarily used as an internal mechanism for the various accessors
/// ([showYearPicker], [showMonthPicker], [showDayPicker]) to signal their
/// required picker type, but is exposed to support direct instantiation of
/// [FlutterDatePicker], where desired.
enum FlutterDatePickerMode {
  /// Show a date picker for choosing a day.
  day,

  /// Show a date picker for choosing a month.
  month,

  /// Show a date picker for choosing a year.
  year,
}

/// Select a year by calling the [YearPicker] class directly.
Future<DateTime> showYearPicker({
  @required BuildContext context,
  @required DateTime initialDate,
  DateTime firstDate,
  DateTime lastDate,
  Widget title,
  EdgeInsetsGeometry titlePadding,
}) async {
  assert(context != null);
  assert(initialDate != null);

  return await showDialog(
    context: context,
    builder: (context) => FlutterDatePicker(
      mode: FlutterDatePickerMode.year,
      selectedDate: initialDate,
      firstDate: firstDate ?? DateTime(0),
      lastDate: lastDate ?? DateTime(9999),
      title: title,
      titlePadding: titlePadding,
    ),
  );
}

/// Select a month by calling the [MonthPicker] class directly.
Future<DateTime> showMonthPicker({
  @required BuildContext context,
  @required DateTime initialDate,
  DateTime firstDate,
  DateTime lastDate,
  Widget title,
  EdgeInsetsGeometry titlePadding,
}) async {
  assert(context != null);
  assert(initialDate != null);

  return await showDialog(
    context: context,
    builder: (context) => FlutterDatePicker(
      mode: FlutterDatePickerMode.month,
      selectedDate: initialDate,
      firstDate: firstDate ?? DateTime(0),
      lastDate: lastDate ?? DateTime(9999),
      title: title,
      titlePadding: titlePadding,
    ),
  );
}

/// Select a day by calling the [DayPicker] class directly.
Future<DateTime> showDayPicker({
  @required BuildContext context,
  @required DateTime initialDate,
  DateTime firstDate,
  DateTime lastDate,
  Widget title,
  EdgeInsetsGeometry titlePadding,
}) async {
  assert(context != null);
  assert(initialDate != null);

  return await showDialog(
    context: context,
    builder: (context) => FlutterDatePicker(
      mode: FlutterDatePickerMode.day,
      selectedDate: initialDate,
      firstDate: firstDate ?? DateTime(0),
      lastDate: lastDate ?? DateTime(9999),
      title: title,
      titlePadding: titlePadding,
    ),
  );
}

/// A non-cascading [SimpleDialog] view of a [FlutterDatePickerMode] picker.
///
/// In general this class does not need to be used directly, and should
/// be primarily accessed through the [showYearPicker], [showMonthPicker]
/// and [showDayPicker] convenience wrappers.
class FlutterDatePicker extends StatefulWidget {
  final FlutterDatePickerMode mode;
  final EdgeInsetsGeometry titlePadding;
  final Widget title;
  final DateTime selectedDate;
  final DateTime firstDate;
  final DateTime lastDate;

  FlutterDatePicker(
      {Key key,
      @required this.mode,
      this.title,
      this.titlePadding,
      @required this.selectedDate,
      @required this.firstDate,
      @required this.lastDate})
      : assert(!firstDate.isAfter(lastDate)),
        assert(
            selectedDate.isAfter(firstDate) && selectedDate.isBefore(lastDate)),
        super(key: key);

  @override
  _FlutterDatePickerState createState() => _FlutterDatePickerState();
}

class _FlutterDatePickerState extends State<FlutterDatePicker> {
  final GlobalKey _pickerKey = GlobalKey();
  DateTime _selectedDate;

  @override
  void initState() {
    super.initState();
    _selectedDate = widget.selectedDate;
  }

  Widget _defaultPickerHeader(BuildContext context) {
    return Container(
        padding: EdgeInsets.all(20.0),
        child: Text("Select a " + describeEnum(widget.mode),
            style: Theme.of(context)
                .textTheme
                .title
                .copyWith(color: Colors.white)),
        color: Theme.of(context).primaryColor);
  }

  @override
  Widget build(BuildContext context) {
    return SimpleDialog(
        titlePadding: widget.titlePadding ?? widget.title == null
            ? EdgeInsets.all(0)
            : EdgeInsets.all(20.0),
        title: widget.title ?? _defaultPickerHeader(context),
        children: <Widget>[
          Container(
            height: 300,
            width: double.maxFinite,
            child: widget.mode == FlutterDatePickerMode.year
                ? YearPicker(
                    key: _pickerKey,
                    firstDate: widget.firstDate,
                    selectedDate: _selectedDate,
                    lastDate: widget.lastDate,
                    onChanged: (DateTime selected) {
                      setState(() {
                        _selectedDate = selected;
                      });
                    },
                  )
                : widget.mode == FlutterDatePickerMode.month
                    ? MonthPicker(
                        key: _pickerKey,
                        firstDate: widget.firstDate,
                        selectedDate: _selectedDate,
                        lastDate: widget.lastDate,
                        onChanged: (DateTime selected) {
                          setState(() {
                            _selectedDate = selected;
                          });
                        },
                      )
                    : DayPicker(
                        key: _pickerKey,
                        displayedMonth: widget.selectedDate,
                        firstDate: widget.firstDate,
                        selectedDate: _selectedDate,
                        currentDate: DateTime.now(),
                        lastDate: widget.lastDate,
                        onChanged: (DateTime selected) {
                          setState(() {
                            _selectedDate = selected;
                          });
                        },
                      ),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.end,
            children: <Widget>[
              FlatButton(
                onPressed: () {
                  Navigator.of(context).pop();
                },
                child: Text('CANCEL',
                    textAlign: TextAlign.end,
                    style: TextStyle(
                        color: Theme.of(context).accentColor,
                        fontWeight: FontWeight.bold)),
              ),
              FlatButton(
                onPressed: () {
                  Navigator.of(context).pop(_selectedDate);
                },
                child: Text('OK',
                    textAlign: TextAlign.end,
                    style: TextStyle(
                        color: Theme.of(context).accentColor,
                        fontWeight: FontWeight.bold)),
              ),
            ],
          ),
        ]);
  }
}
